stdenv, haskell: bonafied GHCJS cross compilation without stdenv.cc

Ericson2314 opened the pull request:
###### Motivation for this change

This commit allows treating GHCJS builds as cross compilation. This is useful for a number of correctness reasons, chiefly getting the right build-time `build-tool-depends` so things like literate READMEs work. 

There is some trickiness to get this to work however. Read the first commit for details. In short, as usual, dependency "splicing" is a terrible idea but one with no better alternative that isn't a huge refactor.

I was very careful to break any interfaces with this e.g. existing native and cross platforms will have the same hashes, and whether `stdenv.cc == null` before, that is still the case. This will allow me to backport this to `19.09` for e.g. reflex-platform and haskell.nix.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @ElvishJerricco @ElvishJerricco @ali-abrar @cdepillabout

alyssais committed:
symlinkJoin: fix cross

(cherry picked from commit 59dbb005556e5bdc0d342ddb1f3801ebcea02413)

Ericson2314 committed:
stdenv: Introduce hasCC attribute

Before, we'd always use `cc = null`, and check for that. The problem is
this breaks for cross compilation to platforms that don't support a C
compiler.

It's a very subtle issue. One might think there is no problem because we
have `stdenvNoCC`, and presumably one would only build derivations that
use that. The problem is that one still wants to use tools at build-time
that are themselves built with a C compiler, and those are gotten via
"splicing". The runtime version of those deps will explode, but the
build time / `buildPackages` versions of those deps will be fine, and
splicing attempts to work this by using `builtins.tryEval` to filter out
any broken "higher priority" packages (runtime is the default and
highest priority) so that both `foo` and `foo.nativeDrv` works.

However, `tryEval` only catches certain evaluation failures (e.g.
exceptions), and not arbitrary failures (such as `cc.attr` when `cc` is
null). This means `tryEval` fails to let us use our build time deps, and
everything comes apart.

The right solution is, as usually, to get rid of splicing. Or, baring
that, to make it so `foo` never works and one has to explicitly do
`foo.*`. But that is a much larger change, and certaily one unsuitable
to be backported to stable.

Given that, we instead make an exception-throwing `cc` attribute, and
create a `hasCC` attribute for those derivations which wish to
condtionally use a C compiler: instead of doing `stdenv.cc or null ==
null` or something similar, one does `stdenv.hasCC`. This allows quering
without "tripping" the exception, while also allowing `tryEval` to work.

No platform without a C compiler is yet wired up by default. That will
be done in a following commit.

Ericson2314 committed:
Add support for cross compiling to `js-ghcjs`

This platform doesn't have a C compiler, and so relies and the changes
in the previous commit to work.

Ericson2314 committed:
haskell genenric-builder: Make the C compiler optional

This is GHCJS, and perhaps other obscure targets.

Ericson2314 committed:
Merge remote-tracking branch 'upstream/master' into ghcjs-cross-without-cc

ElvishJerricco commented on pkgs/development/interpreters/perl/default.nix:
```diff
@@ -144,7 +144,7 @@ let
           --replace "${libcInc}" /no-such-path \
           --replace "${
-              if stdenv.cc.cc or null != null then stdenv.cc.cc else "/no-such-path"
+              if stdenv.hasCC then stdenv.cc.cc else "/no-such-path"
```
Why not just `optionalString` the whole `--replace` argument?

Ericson2314 replied:
Maybe eventually, but trying to avoid a mass rebuild for now :)

Ericson2314 committed:
Fix lib tests

js-ghcjs didn't fit in an existing categor.

Ericson2314 committed:
Fix lib tests

js-ghcjs didn't fit in an existing categor.

Ericson2314 committed:
Merge branch 'ghcjs-cross-without-cc-19.09' into ghcjs-cross-without-cc

matthewbauer commented on pkgs/stdenv/cross/default.nix:
```diff
@@ -51,12 +51,18 @@ in lib.init bootStages ++ [
+             # `null` or skipping the attribute would cause an eval failure
+             # `tryEval` wouldn't catch, wrecking accessing previous stages
+             # when there is a C compiler and everything should be fine.
+             then throw "no C compile provided for this platform"
```
```suggestion
             then throw "no C compiler provided for this platform"
```

Ericson2314 replied:
Thanks!

Ericson2314 committed:
Merge remote-tracking branch 'obsidian/ghcjs-cross-without-cc-19.09' into ghcjs-cross-without-cc

matthewbauer commented:
Can you add at least one test for release-cross.nix as well?

Ericson2314 commented:
Great idea

Ericson2314 committed:
release-cross.nix: Add GHCJS-built hello

Ericson2314 committed:
Merge branch 'ghcjs-cross-without-cc-19.09' into ghcjs-cross-without-cc

Ericson2314 committed:
ghcWithHoogle: Fix for cross

Use `buildPackages.stdenv.mkDerivation` because we are making a shell
script to start hoogle on the build platform.

Ericson2314 commented:
I added it, but it won't work yet (at least on 19.09) because ghcjs itself doesn't build. I think that's fine; I rather upstream whatever reflex-platform does to fix it after this change.

Ericson2314 committed:
ghcjs: get socket.io from pkgsHostHost

This is a bit dubvious, but the alternative of making nodejs a
nativeBuildInput for node packages is worse. In general the cross story
for interpreted languages is murky, and this fits that pattern.

Commit:
treewide: Check `stdenv.isi686` before checking `stdenv.cc.isGNU`

This makes us a bit more robust to various splicing nastiness. May splicing
someday go so we don't have to resort to such hacks.

Ericson2314 committed:
Merge branch 'ghcjs-cross-without-cc-common' into ghcjs-cross-without-cc

Ericson2314 committed:
arrow-cpp, llvmPackage_9: Check stdenv.isi686 before stdenv.cc.isGNU

See c3c245dcda8cc54055bca7fe08c59dd1d3a41408

Ericson2314 committed:
Merge branch 'ghcjs-cross-without-cc-common' into ghcjs-cross-without-cc

Ericson2314 committed:
haskell generic-builder: Fix `--with-ghc` flag for GHCJS as cross

Otherwise it passes `--with-ghc=ghc`, and we do the wrong thing.

Ericson2314 committed:
Merge branch 'ghcjs-cross-without-cc-common' into ghcjs-cross-without-cc

Ericson2314 committed:
haskell generic-builder: Hack so CPP without a CC works

Ericson2314 committed:
Merge branch 'ghcjs-cross-without-cc-common' into ghcjs-cross-without-cc

matthewbauer commented on pkgs/stdenv/booter.nix:
```diff
@@ -121,9 +121,12 @@ stageFuns: let
+        if buildPackages.stdenv.cc.isClang or false
+        then buildPackages.clang
+        else buildPackages.gcc
+      else buildPackages.stdenv.cc;
```
is this always null?

matthewbauer replied:
otherwise it needs some comments

matthewbauer commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -156,7 +160,9 @@ let
     "--libsubdir=\\$abi/\\$libname"
     (optionalString enableSeparateDataOutput "--datadir=$data/share/${ghc.name}")
     (optionalString enableSeparateDocOutput "--docdir=${docdir "$doc"}")
+  ] ++ optionals stdenv.hasCC [
```
This is a little weird. Isn't --with-gcc always needed for GHC?

Ericson2314 replied:
I suppose but those "cross flags" do that. We should really just combine the list of flags and make cross and native more similar.

Ericson2314 committed:
pkgs/stdenv/booter.nix: Add comment explaining hasCC trickery

Ericson2314 committed:
Merge remote-tracking branch 'obsidian/ghcjs-cross-without-cc-common' into ghcjs-cross-without-cc
